actions-nightly-trigger/action.yaml

```diff
@@ -0,0 +1,188 @@
+name: 'Smart Nightly Trigger'
+description: 'Intelligently decides whether to trigger nightly workflows based on commits and schedule'
+
+# Required permissions for the calling workflow:
+#   contents: read      # For checking out the repository and reading git history
+#   actions: read       # For querying workflow run status and jobs via GitHub API
+#
+# Note: The access_token input must be a Personal Access Token (PAT) or GitHub App token
+#       with 'repo' and 'workflow' scopes to trigger workflows and retry failed jobs.
+#       The default GITHUB_TOKEN cannot trigger workflows due to GitHub security restrictions.
+
+inputs:
+  branch:
+    description: 'Branch to check and trigger workflow for'
+    required: true
+  workflow:
+    description: 'Workflow to trigger (e.g., nightly.yaml)'
+    required: true
+    default: 'nightly.yaml'
+  force-trigger:
+    description: 'Force trigger regardless of conditions'
+    required: false
+    default: 'false'
+  max-retries:
+    description: 'Maximum number of retries for failed jobs (0 to disable retries)'
+    required: false
+    default: '3'
+  access_token:
+    description: 'Access token for triggering workflows (required)'
+    required: true
+
+runs:
+  using: composite
+  steps:
+    - name: Checkout repository
+      uses: actions/checkout@v5
+      shell: bash
+      with:
+        ref: ${{ inputs.branch }}
+        fetch-depth: 0
```

Comment:
this actually fetches the whole history.
to retrict fetch size, should be `fetch-depth: 1`

(we only need the last commit, so not an issue to get fetch-depth 1)